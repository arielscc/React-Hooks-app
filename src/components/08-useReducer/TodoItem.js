import React from 'react';
import PropTypes from 'prop-types';

const TodoItem = React.memo(({ todo, index, handleToggle, handleDelete }) => {
  return (
    <li
      key={todo.id}
      onClick={() => handleToggle(todo.id)}
      className={`flex justify-between items-center mt-2 w-1/2 p-3 rounded-md cursor-pointer box-content  ${
        index % 2 === 0 && 'bg-gray-200'
      } ${
        todo.done &&
        'line-through bg-gradient-to-r from-indigo-500 to-indigo-300'
      }`}
    >
      <p>{todo.todo}</p>
      <button
        onClick={() => handleDelete(todo.id)}
        className="py-1 px-2 rounded-md bg-red-600 font-bold text-white focus:outline-none"
      >
        <svg
          className="w-6 h-6"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={2}
            d="M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1 1v3M4 7h16"
          />
        </svg>
      </button>
    </li>
  );
});

TodoItem.propTypes = {
  todo: PropTypes.object.isRequired,
  index: PropTypes.number.isRequired,
  handleToggle: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired,
};

export default TodoItem;
