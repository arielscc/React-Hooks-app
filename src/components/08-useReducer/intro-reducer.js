const initialState = [
  {
    id: 1,
    todo: 'Lavar la ropa',
    done: false,
  },
];

const todoReducer = (state = initialState, action) => {
  if (action?.type === 'guardar') {
    return [...state, action.payload];
  }
  return state;
};

let todos = todoReducer();

console.log(todos);

const newTodo = {
  id: 2,
  todo: 'Comprar pan',
  done: false,
};

const todoAction = {
  type: 'guardar',
  payload: newTodo,
};

todos = todoReducer(todos, todoAction);

console.log(todos);
