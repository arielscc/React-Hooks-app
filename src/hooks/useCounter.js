import { useState } from 'react';

export const useCounter = (initialState = 10) => {
  const [counter, setCounter] = useState(initialState);

  const increment = (e, value = 1) => {
    return setCounter(counter + value);
  };
  const decrement = (e, value = 1) => {
    return setCounter(counter - value);
  };

  const reset = () => {
    return setCounter(initialState);
  };

  return {
    counter,
    increment,
    decrement,
    reset,
  };
};
