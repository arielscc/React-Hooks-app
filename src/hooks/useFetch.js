import { useState, useEffect, useRef } from 'react';

export const useFetch = (url) => {
  const initialState = {
    data: null,
    loading: true,
    error: null,
  };
  //
  const isMounted = useRef(true);
  useEffect(() => {
    return () => {
      isMounted.current = false;
    };
  }, []);

  const [state, setState] = useState(initialState);

  useEffect(() => {
    fetch(url)
      .then((resp) => resp.json())
      .then((data) => {
        if (isMounted.current) {
          setState({
            loading: false,
            error: null,
            data,
          });
        }
      });
    setState(initialState);
  }, [url]);

  return state;
};
