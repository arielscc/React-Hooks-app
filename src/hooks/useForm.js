import { useState } from 'react';

export const useForm = (initialState = {}) => {
  const [values, setValues] = useState(initialState);
  console.log('useform');
  const handleInputChange = ({ target }) => {
    setValues({
      ...values,
      [target.name]: target.value,
    });
  };

  const resetValues = () => setValues(initialState);

  return [values, handleInputChange, resetValues];
};
