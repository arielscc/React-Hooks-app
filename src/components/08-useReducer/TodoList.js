import React from 'react';
import PropTypes from 'prop-types';
import TodoItem from './TodoItem';

const TodoList = React.memo(({ todos, handleToggle, handleDelete }) => {
  return (
    <ul>
      {todos.map((todo, i) => (
        <TodoItem
          key={todo.id}
          todo={todo}
          index={i}
          handleToggle={handleToggle}
          handleDelete={handleDelete}
        />
      ))}
    </ul>
  );
});

TodoList.propTypes = {
  todos: PropTypes.array.isRequired,
  handleToggle: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired,
};

export default TodoList;
