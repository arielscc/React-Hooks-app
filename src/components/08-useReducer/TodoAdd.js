import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAdd }) => {
  const [{ description }, handleInputChange, resetValues] = useForm({
    description: '',
  });

  const handleSubmit = (e) => {
    e.preventDefault();
    if (description.trim().length <= 2) return;

    const newTodo = {
      id: new Date().getTime(),
      todo: description,
      done: false,
    };

    handleAdd(newTodo);
    resetValues();
  };

  return (
    <form onSubmit={handleSubmit} className="my-4">
      <input
        type="text"
        name="description"
        className="py-2 px-4 w-1/2 border rounded-md focus:outline-none block"
        autoComplete="off"
        value={description}
        onChange={handleInputChange}
      />
      <button
        type="submit"
        className="mt-2 py-1 px-3 rounded-md bg-teal-600 font-bold text-white focus:outline-none"
      >
        Add todo
      </button>
    </form>
  );
};
