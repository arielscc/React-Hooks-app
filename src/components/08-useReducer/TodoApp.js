import React, { useCallback, useEffect, useReducer } from 'react';
import { TodoAdd } from './TodoAdd';
import TodoList from './TodoList';
import { todoReducer } from './todoReducer';

const init = () => {
  return JSON.parse(localStorage.getItem('todos')) || [];
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  // console.log(todos);
  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const handleDelete = useCallback(
    (todoId) => {
      const action = {
        type: 'delete',
        payload: todoId,
      };

      dispatch(action);
    },
    [dispatch]
  );

  const handleToggle = useCallback((todoId) => {
    const action = {
      type: 'toggle',
      payload: todoId,
    };
    dispatch(action);
  }, []);

  const handleAdd = (newTodo) => {
    dispatch({
      type: 'add',
      payload: newTodo,
    });
  };

  return (
    <div className="p-20">
      <p className="flex gap-2 items-center text-gray-700 mb-4">
        <svg
          className="w-6 h-6 "
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={2}
            d="M4 6a2 2 0 012-2h2a2 2 0 012 2v2a2 2 0 01-2 2H6a2 2 0 01-2-2V6zM14 6a2 2 0 012-2h2a2 2 0 012 2v2a2 2 0 01-2 2h-2a2 2 0 01-2-2V6zM4 16a2 2 0 012-2h2a2 2 0 012 2v2a2 2 0 01-2 2H6a2 2 0 01-2-2v-2zM14 16a2 2 0 012-2h2a2 2 0 012 2v2a2 2 0 01-2 2h-2a2 2 0 01-2-2v-2z"
          />
        </svg>
        <span className="text-lg font-bold">todo app {todos.length}</span>
      </p>
      <hr className="w-1/2" />

      <TodoAdd handleAdd={handleAdd} />

      <TodoList
        todos={todos}
        handleToggle={handleToggle}
        handleDelete={handleDelete}
      />
    </div>
  );
};
