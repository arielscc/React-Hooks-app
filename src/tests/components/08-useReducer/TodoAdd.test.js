import React from 'react';
import { shallow } from 'enzyme';
import { TodoAdd } from '../../../components/08-useReducer/TodoAdd';
import { useForm } from '../../../hooks/useForm';

const handleAdd = jest.fn();

describe('Pruebas en el componente <TodoAdd/>', () => {
  const wrapper = shallow(<TodoAdd handleAdd={handleAdd} />);

  test('Debe de mostrarse correctamente el componente <TodoAdd/>', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('No debe de llamar el habdleAddTodo cuando el value del input no existe', () => {
    // forma 1 de simular el envio de formulario
    const formSubmit = wrapper.find('form').prop('onSubmit');
    formSubmit({ preventDefault() {} });

    expect(handleAdd).toHaveBeenCalledTimes(0);
    // console.log(formSubmit);
  });

  test('Debe de llamar la funcion handleAdd', () => {
    const value = 'Aprender Testing';
    wrapper.find('input').simulate('change', {
      target: {
        name: 'description',
        value,
      },
    });
    const formSubmit = wrapper.find('form').prop('onSubmit');
    formSubmit({ preventDefault() {} });

    // forma 2 de simular el envio de formulario
    // wrapper.find('form').simulate('submit', { preventDefault() {} });

    expect(handleAdd).toHaveBeenCalledTimes(1);
    expect(handleAdd).toHaveBeenCalledWith(expect.any(Object));
    expect(handleAdd).toHaveBeenCalledWith({
      id: expect.any(Number),
      todo: value,
      done: false,
    });
    expect(wrapper.find('input').prop('value')).toBe('');
  });
});
