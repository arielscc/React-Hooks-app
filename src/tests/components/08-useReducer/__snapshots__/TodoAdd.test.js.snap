// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <TodoAdd/> Debe de mostrarse correctamente el componente <TodoAdd/> 1`] = `
<form
  className="my-4"
  onSubmit={[Function]}
>
  <input
    autoComplete="off"
    className="py-2 px-4 w-1/2 border rounded-md focus:outline-none block"
    name="description"
    onChange={[Function]}
    type="text"
    value=""
  />
  <button
    className="mt-2 py-1 px-3 rounded-md bg-teal-600 font-bold text-white focus:outline-none"
    type="submit"
  >
    Add todo
  </button>
</form>
`;
